src/library_tty.js

```diff
@@ -124,36 +123,36 @@ mergeInto(LibraryManager.library, {
         }
         return tty.input.shift();
       },
-      flush: function(tty) {
-        if (tty.output && tty.output.length > 0) {
-          Module['print'](tty.output.join(''));
-          tty.output = [];
-        }
-      },
       put_char: function(tty, val) {
         if (val === null || val === {{{ charCode('\n') }}}) {
-          Module['print'](tty.output.join(''));
+          Module['print'](UTF8ArrayToString(tty.output, 0));
           tty.output = [];
         } else {
-          tty.output.push(TTY.utf8.processCChar(val));
+          if (val != 0) tty.output.push(val); // val == 0 would cut text output off in the middle.
+        }
+      },
+      flush: function(tty) {
+        if (tty.output && tty.output.length > 0) {
+          Module['print'](UTF8ArrayToString(tty.output, 0));
```

Comment:
Ran a test on that with the allocation logger, and it shows the following:

	"4942: (root)" Scratchpad/1:126:0
	"   448: PreviewController.prototype.handleEvent  resource://app/modules/WindowsPreviewPerTab.jsm:361:14" Scratchpad/1:126:0
	"   448: PreviewController.prototype.handleEvent  resource://app/modules/WindowsPreviewPerTab.jsm:364:20" Scratchpad/1:126:0
	"   2100: Browser_mainLoop_runner  file:///C:/code/emsdk/emscripten/incoming/a.js:1585:2" Scratchpad/1:126:0
	"      1575: Browser.mainLoop.runIter  file:///C:/code/emsdk/emscripten/incoming/a.js:5441:4" Scratchpad/1:126:0
	"         1575: Browser_mainLoop_runner/<  file:///C:/code/emsdk/emscripten/incoming/a.js:1589:4" Scratchpad/1:126:0
	"            1575: Runtime.dynCall  file:///C:/code/emsdk/emscripten/incoming/a.js:193:10" Scratchpad/1:126:0
	"               1575: dynCall_v  file:///C:/code/emsdk/emscripten/incoming/a.js:8033:1" Scratchpad/1:126:0
	"                  1575: __Z4tickv  file:///C:/code/emsdk/emscripten/incoming/a.js:8026:1" Scratchpad/1:126:0
	"                     1575: _puts  file:///C:/code/emsdk/emscripten/incoming/a.js:5381:19" Scratchpad/1:126:0
	"                        1575: _fputc  file:///C:/code/emsdk/emscripten/incoming/a.js:5366:11" Scratchpad/1:126:0
	"                           1575: _write  file:///C:/code/emsdk/emscripten/incoming/a.js:4949:9" Scratchpad/1:126:0
	"                              1575: FS.write  file:///C:/code/emsdk/emscripten/incoming/a.js:3827:21" Scratchpad/1:126:0
	"                                 1575: TTY.stream_ops.write  file:///C:/code/emsdk/emscripten/incoming/a.js:1899:5" Scratchpad/1:126:0
	"                                    1574: TTY.default_tty_ops.put_char  file:///C:/code/emsdk/emscripten/incoming/a.js:1942:0" Scratchpad/1:126:0
	"                                       525: Module.print</<  file:///C:/code/emsdk/emscripten/incoming/a.html:1226:0" Scratchpad/1:126:0
	"                                       525: Module.print</<  file:///C:/code/emsdk/emscripten/incoming/a.html:1227:19" Scratchpad/1:126:0
	"                                    1: TTY.default_tty_ops.put_char  file:///C:/code/emsdk/emscripten/incoming/a.js:1943:4"

this was tested on the following application:

    void tick() {
      printf("Hello there, this is a message! Hello there, this is a message! Hello there, this is a message! Hello there, this is a message! Hello there, this is a message! Hello there, this is a message!\n");
    }

    int main() {
      emscripten_set_main_loop(tick, 0, 0);
    }

so I think we are good here.